data/trie/patriciaMerkleTrie.go

```diff
@@ -400,7 +428,45 @@ func (tr *patriciaMerkleTrie) ResetOldHashes() [][]byte {
 	return oldHashes
 }
 
-// AppendToOldHashes appends the given hashes to the trie's oldHashes variable
-func (tr *patriciaMerkleTrie) AppendToOldHashes(hashes [][]byte) {
-	tr.oldHashes = append(tr.oldHashes, hashes...)
+// Snapshot creates a new database in which the current state of the trie is saved.
+// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
+func (tr *patriciaMerkleTrie) Snapshot() error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	err := tr.root.commit(false, 0, tr.db, tr.marshalizer, tr.hasher)
+	if err != nil {
+		return err
+	}
+
+	db, err := storageUnit.NewDB(
+		storageUnit.DBType(tr.snapshotDbCfg.Type),
+		path.Join(tr.snapshotDbCfg.FilePath, strconv.Itoa(tr.snapshotId)),
+		tr.snapshotDbCfg.BatchDelaySeconds,
+		tr.snapshotDbCfg.MaxBatchSize,
+		tr.snapshotDbCfg.MaxOpenFiles,
+	)
+	if err != nil {
+		return err
+	}
+	tr.snapshotId++
+
+	err = tr.root.commit(true, 0, db, tr.marshalizer, tr.hasher)
+	if err != nil {
+		return err
+	}
+
+	tr.snapshots = append(tr.snapshots, db)
+
+	if len(tr.snapshots) > maxSnapshots {
+		dbUniqueId := strconv.Itoa(tr.snapshotId - len(tr.snapshots))
```

Comment:
`fmt.Sprintf()` ?

Reply:
`fmt.Sprintf()` is slower than `strconv`